applications/StructuralMechanicsApplication/custom_elements/timoshenko_beam_element_2D2N.cpp

```diff
@@ -0,0 +1,1018 @@
+        rN.resize(4, false);
+    rN[0] = (3.0 * xi * xi - 3.0) / (2.0 * one_plus_phi * Length);
+    rN[1] = (xi - 1.0) * (1.0 + 3.0 * xi - 2.0 * Phi) / (4.0 * one_plus_phi);
+    rN[2] = (3.0 - 3.0 * xi * xi) / (2 * one_plus_phi * Length);
+    rN[3] = (1.0 + xi) * (3.0 * xi - 1.0 + 2.0 * Phi) / (4.0 * one_plus_phi);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::GetFirstDerivativesNThetaShapeFunctionsValues(
+    VectorType& rN,
+    const double Length,
+    const double Phi,
+    const double xi
+    )
+{
+    const double one_plus_phi = 1.0 + Phi;
+    if (rN.size() != 4)
+        rN.resize(4, false);
+    rN[0] = 3.0 * xi / (one_plus_phi * Length);
+    rN[1] = (-0.5 * Phi + 1.5 * xi - 0.5) / (one_plus_phi);
+    rN[2] = (-3.0 * xi) / (Length * Phi + Length);
+    rN[3] = (0.5 * Phi + 1.5 * xi + 0.5) / (one_plus_phi);
+    rN *= (2.0 / Length);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::GetNu0ShapeFunctionsValues(
+    VectorType& rN,
+    const double Length,
+    const double Phi,
+    const double xi
+    )
+{
+    if (rN.size() != 2)
+        rN.resize(2, false);
+    rN[0] = 0.5 * (1.0 - xi);
+    rN[1] = 0.5 * (1.0 + xi);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::GetFirstDerivativesNu0ShapeFunctionsValues(
+    VectorType& rN,
+    const double Length,
+    const double Phi,
+    const double xi
+    )
+{
+    if (rN.size() != 2)
+        rN.resize(2, false);
+    const double inverse_l = 1.0 / Length;
+    rN[0] = -inverse_l;
+    rN[1] =  inverse_l;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::GetNodalValuesVector(VectorType& rNodalValues)
+{
+    if (rNodalValues.size() != 6)
+        rNodalValues.resize(6, false);
+    const auto &r_geom = GetGeometry();
+
+    const double angle = StructuralMechanicsElementUtilities::
+        GetReferenceRotationAngle2D2NBeam(GetGeometry());
+
+    if (std::abs(angle) > std::numeric_limits<double>::epsilon()) {
+        BoundedMatrix<double, 3, 3> T;
+        BoundedVector<double, 6> global_values;
+        BoundedMatrix<double, 6, 6> global_size_T;
+        StructuralMechanicsElementUtilities::BuildRotationMatrixFor2D2NBeam(T, angle);
+        StructuralMechanicsElementUtilities::BuildElementSizeRotationMatrixFor2D2NBeam(T, global_size_T);
+
+        global_values[0] = r_geom[0].FastGetSolutionStepValue(DISPLACEMENT_X);
```

Comment:
here you are accessing twice the database. it would be faster to do

~~~cpp
   const auto& d0 = r_geom[0].FastGetSolutionStepValue(DISPLACEMEN);
global_values[0] = v[0];
        global_values[1] = v[1];
        global_values[2] = r_geom[0].FastGetSolutionStepValue(ROTATION_Z);
~~~

(same with the second node)